syntax = "proto3";

package example.v1;

import "google/protobuf/timestamp.proto";

// The greeting service definition
service GreeterService {
  // Sends a greeting
  rpc SayHello(HelloRequest) returns (HelloResponse);

  // Sends multiple greetings (server streaming)
  rpc SayHelloStream(HelloRequest) returns (stream HelloResponse);

  // Receives multiple requests (client streaming)
  rpc SayHelloClientStream(stream HelloRequest) returns (HelloResponse);

  // Bidirectional streaming
  rpc SayHelloBidi(stream HelloRequest) returns (stream HelloResponse);
}

// The request message containing the user's name
message HelloRequest {
  string name = 1;
  string language = 2; // Language preference for greeting
  google.protobuf.Timestamp timestamp = 3;
}

// The response message containing the greeting
message HelloResponse {
  string message = 1;
  string language = 2;
  google.protobuf.Timestamp server_timestamp = 3;
  int32 response_count = 4; // For streaming responses
}

